### PR TITLE
Remove staking&session from impl details of the pallet elections

### PR DESCRIPTION
**Pull request by kostekIV**
## Type of change

Please delete options that are not relevant.
- New feature (non-breaking change which adds functionality)

# Checklist:

- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Infrastructure updated accordingly
- [ ] I have made corresponding changes to the documentation
- [ ] New documentation created
- [ ] Bump `spec_version` and `transaction_version` if relevant
- [ ] Bump `aleph-client` version if relevant

### CODE REVIEW

**Review by Yithis — APPROVED** [review]
LGTM. Very good changes, code is way easier to read after those

File: pallets/elections/src/lib.rs [hunk 1]
```diff
@@ -56,6 +58,7 @@ pub mod pallet {
 
     #[pallet::config]
     pub trait Config: frame_system::Config {
+        type EraInfoProvider: EraInfoProvider;
```
Comment (Marcin-Radecki) [inline pallets/elections/src/lib.rs:61]: Shall we add some comments to new types and maybe existing ones? As pallet keeps growing. It seems it is standard Substarte way to comment `type`s for `pallet::Config`

File: pallets/elections/src/traits.rs [hunk 2]
```diff
@@ -0,0 +1,76 @@
+use frame_support::{pallet_prelude::Get, traits::Currency};
+use sp_std::vec::Vec;
+
+pub type SessionId = u32;
+pub type EraId = u32;
+
+pub trait SessionInfoProvider<T: frame_system::Config> {
```
Comment (Marcin-Radecki) [inline pallets/elections/src/traits.rs:7]: Could we please add API comments here as well, as those are `pub trait`s

File: pallets/elections/src/traits.rs [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    fn current_session() -> SessionId;
+    fn current_committee() -> Vec<T::AccountId>;
+}
+
+impl<T> SessionInfoProvider<T> for pallet_session::Pallet<T>
+where
+    T: pallet_session::Config,
+    T::ValidatorId: Into<T::AccountId>,
+{
+    fn current_session() -> SessionId {
```
Comment (Marcin-Radecki) [inline pallets/elections/src/traits.rs:17]: Nit: `current_session_index()` ?

File: pallets/elections/src/traits.rs [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    fn current_committee() -> Vec<T::AccountId> {
+        pallet_session::Validators::<T>::get()
+            .into_iter()
+            .map(|a| a.into())
+            .collect()
+    }
+}
+
+pub trait ValidatorRewardsHandler<T: frame_system::Config> {
+    fn all_era_validators(era: EraId) -> Vec<T::AccountId>;
```
Comment (Marcin-Radecki) [inline pallets/elections/src/traits.rs:30]: Shall we comment here that order of validators is not upfront determined, yet deterministic for this same runtime?

File: pallets/elections/src/traits.rs [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+where
+    T: pallet_staking::Config,
+    <T::Currency as Currency<T::AccountId>>::Balance: Into<u128>,
+{
+    fn all_era_validators(era: EraId) -> Vec<T::AccountId> {
+        pallet_staking::ErasStakers::<T>::iter_key_prefix(era).collect()
+    }
+
+    fn validator_totals(era: EraId) -> Vec<(T::AccountId, u128)> {
+        pallet_staking::ErasStakers::<T>::iter_prefix(era)
```
Comment (Marcin-Radecki) [inline pallets/elections/src/traits.rs:45]: This is potentially the most 'heavy' read in all newly added pallet elections storage reads?

File: pallets/elections/src/traits.rs [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+impl<T> EraInfoProvider for pallet_staking::Pallet<T>
+where
+    T: pallet_staking::Config,
+{
+    fn current_era() -> Option<EraId> {
+        pallet_staking::ActiveEra::<T>::get().map(|ae| ae.index)
+    }
+
+    fn era_start(era: EraId) -> Option<SessionId> {
```
Comment (Marcin-Radecki) [inline pallets/elections/src/traits.rs:69]: Nit: I think original naming with session index adds value here, so `era_start_session_index`.

File: pallets/elections/src/impls.rs [hunk 7]
```diff
@@ -157,15 +155,15 @@ impl<T> Pallet<T>
     // Choose a subset of all the validators for current era that contains all the
     // reserved nodes. Non reserved ones are chosen in consecutive batches for every session
     fn rotate_committee() -> Option<Vec<T::AccountId>> {
-        let current_era = match pallet_staking::ActiveEra::<T>::get() {
-            Some(ae) if ae.index > 0 => ae.index,
+        let current_era = match T::EraInfoProvider::current_era() {
```
Comment (Marcin-Radecki) [inline pallets/elections/src/impls.rs:158]: Nit: I think `active_era` name would make sense as well.

File: pallets/elections/src/traits.rs (hunk 2 above)
Comment (kostekIV) [reply pallets/elections/src/traits.rs:5]: added

File: pallets/elections/src/traits.rs (hunk 5 above)
Comment (kostekIV) [reply pallets/elections/src/traits.rs:48]: yes

File: pallets/elections/src/traits.rs (hunk 6 above)
Comment (kostekIV) [reply pallets/elections/src/traits.rs]: done

File: pallets/elections/src/impls.rs (hunk 7 above)
Comment (kostekIV) [reply pallets/elections/src/impls.rs]: done

File: pallets/elections/src/traits.rs (hunk 3 above)
Comment (kostekIV) [reply pallets/elections/src/traits.rs]: done

File: pallets/elections/src/traits.rs (hunk 4 above)
Comment (Marcin-Radecki) [reply pallets/elections/src/traits.rs]: Right, it is more like public knowledge.

**Review by Marcin-Radecki — APPROVED** [review]

**Pull request merged**